chain-crypto/src/digest.rs

```diff
@@ -242,3 +246,80 @@ impl<H: DigestAlg> fmt::Debug for Digest<H> {
         f.write_str(")")
     }
 }
+
+impl<H: DigestAlg> Digest<H> {
+    /// Get the digest of a slice of data
+    pub fn digest(slice: &[u8]) -> Self {
+        let mut ctx = Context::new();
+        ctx.append_data(slice);
+        ctx.finalize()
+    }
+}
+
+use std::marker::PhantomData;
+
+/// A typed version of Digest
+pub struct DigestOf<H: DigestAlg, T> {
+    inner: Digest<H>,
+    marker: PhantomData<T>,
+}
+
+impl<H: DigestAlg, T> Clone for DigestOf<H, T> {
+    fn clone(&self) -> Self {
+        DigestOf {
+            inner: self.inner.clone(),
+            marker: self.marker.clone(),
+        }
+    }
+}
+
+impl<H: DigestAlg, T> From<DigestOf<H, T>> for Digest<H> {
+    fn from(d: DigestOf<H, T>) -> Self {
+        d.inner
+    }
+}
+
+impl<H: DigestAlg, T> From<Digest<H>> for DigestOf<H, T> {
```

Comment:
What is this transformation for? Presumably `DigestOf` is used to calculate digest of a particular type of value, but here it can also be obtained from an arbitrary digest?

Reply:
a `DigestOf<T>` is a `Digest` but tied to a given type. You state this is the digest of a `T`.
For example, the FragmentId is a `DigestOf<Fragment>`. Here we allow users to tag any arbitrary `Digest` with any arbitrary type `T`.

Reply:
OK, but then I don't see a point in this complication.